relentless/optimize/method.py

```diff
@@ -0,0 +1,83 @@
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        pass
+
+class SteepestDescent(Optimizer):
+    def __init__(self, alpha, max_iter, abs_tol):
+        super().__init__()
+        self._alpha = alpha
+        self._max_iter = max_iter
+        self._abs_tol = abs_tol
+
+    def optimize(self, objective):
+        dvars = objective.design_variables()
+        if len(dvars) == 0:
+            return False
+
+        iter_num = 0
+        converged = False
+        while not converged and iter_num < self.max_iter:
+            res = objective.compute()
+            converged = self._has_converged(dvars, res)
+            for x in dvars:
+                x.value -= self.alpha*res.gradient(x)
```

Comment:
I was thinking about this again, and I'm not sure whether or not you should apply the update if the algorithm is determined to have converged. The distinction probably doesn't matter too much, but we should make sure we're in agreement. Let's discuss at our meeting.